quil-rs/src/instruction/extern_call.rs

```diff
@@ -0,0 +1,2060 @@
+use nom_locate::LocatedSpan;
+use num_complex::Complex64;
+
+use crate::{
+    expression::format_complex,
+    hash::hash_f64,
+    parser::{lex, InternalParseError, ParserErrorKind, ParserInput},
+    program::{disallow_leftover, MemoryRegion, SyntaxError},
+    quil::Quil,
+    validation::identifier::{validate_user_identifier, IdentifierValidationError},
+};
+
+use super::{MemoryReference, ScalarType, Vector};
+
+/// A parameter type within an extern signature.
+#[derive(Clone, Debug, PartialEq, Hash, Eq)]
+pub enum ExternParameterType {
+    Scalar(ScalarType),
+    FixedLengthVector(Vector),
+    VariableLengthVector(ScalarType),
+}
+
+impl Quil for ExternParameterType {
+    fn write(
+        &self,
+        f: &mut impl std::fmt::Write,
+        fall_back_to_debug: bool,
+    ) -> crate::quil::ToQuilResult<()> {
+        match self {
+            ExternParameterType::Scalar(value) => value.write(f, fall_back_to_debug),
+            ExternParameterType::FixedLengthVector(value) => value.write(f, fall_back_to_debug),
+            ExternParameterType::VariableLengthVector(value) => {
+                value.write(f, fall_back_to_debug)?;
+                write!(f, "[]").map_err(Into::into)
+            }
+        }
+    }
+}
+
+/// An extern parameter with a name, mutability, and data type.
+#[derive(Clone, Debug, PartialEq, Eq, Hash)]
+pub struct ExternParameter {
+    /// The name of the parameter. This must be a valid user identifier.
+    pub name: String,
+    /// Whether the parameter is mutable.
+    pub mutable: bool,
+    /// The data type of the parameter.
+    pub data_type: ExternParameterType,
+}
+
+impl ExternParameter {
+    /// Create a new extern parameter.
+    pub fn new(name: String, mutable: bool, data_type: ExternParameterType) -> Self {
+        Self {
+            name,
+            mutable,
+            data_type,
+        }
+    }
+}
+
+impl Quil for ExternParameter {
+    fn write(
+        &self,
+        writer: &mut impl std::fmt::Write,
+        fall_back_to_debug: bool,
+    ) -> Result<(), crate::quil::ToQuilError> {
+        write!(writer, "{} : ", self.name)?;
+        if self.mutable {
+            write!(writer, "mut ")?;
+        }
+        self.data_type.write(writer, fall_back_to_debug)
+    }
+}
+
+/// An extern signature with a return type and parameters.
+#[derive(Clone, Debug, PartialEq, Eq, Hash)]
+pub struct ExternSignature {
+    /// The return type of the extern signature, if any.
+    pub return_type: Option<ScalarType>,
+    /// The parameters of the extern signature.
+    pub parameters: Vec<ExternParameter>,
+}
+
+impl ExternSignature {
+    /// Create a new extern signature.
+    pub fn new(return_type: Option<ScalarType>, parameters: Vec<ExternParameter>) -> Self {
+        Self {
+            return_type,
+            parameters,
+        }
+    }
+
+    fn has_return_or_parameters(&self) -> bool {
+        self.return_type.is_some() || !self.parameters.is_empty()
+    }
```

Comment:
Maybe better to make this required in the (fallible) constructor and make the fields private?

Reply:
Yes on the fallible constructor. Private fields present a couple issues:

* The Python package uses this `py_wrap_data_struct!` macro that requires the fields to be public.
* Most other instructions in quil-rs have fully public fields.

There's not really precedent to create an intermediate in `quil-py` with public fields that then converts to a `quil-rs` `Call`, so I hesitate to introduce one here.

Reply:
Ok there is `rigetti_pyo3::py_wrap_type!` to handle this case. It's a bit more verbose, but otherwise fine to support this.